### PR TITLE
[CodeCompletion] Add decl context dependent 'Flair' to decl keywords 

### PR DESCRIPTION
**Pull request by rintaro**
Based on #37563 

Introduce `CommonKeywordAtCurrentPosition` and `RareKeywordAtCurrentPosition`

* `super` in a  overriding decl is "common".
* type decl introducers (e.g. `struct`, `enum`) at top-level in library files are "common"
* type decl introducers in `protocol` are invalid, hence "rare"
* top-level only decl introducer (e.g. `import`, `extension`) are invalid at non-top-level, hence "rare"
* nested types in function bodies are "rare"
* member only decls (e.g. `subscript`, `deinit`) are invalid in function body, hence "rare"
* some modifiers (e.g. `public`, `private`, `override`) are invalid for local decls, hence "rare"

rdar://77934651

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test macOS

File: lib/IDE/CodeCompletion.cpp [hunk 1]
```diff
@@ -1822,6 +1830,39 @@ static bool canDeclContextHandleAsync(const DeclContext *DC) {
...
+    if (CodeCompletionExpr *CCExpr = dyn_cast<CodeCompletionExpr>(expr)) {
+      if (CCExpr->getBase() == nullptr)
+        return true;
+    }
+  }
+
+  return false;
+}
+
+static bool isCompletionDeclContextLocalContext(DeclContext *DC) {
```
[the remark this answers is not in the document]
Comment (rintaro) [reply lib/IDE/CodeCompletion.cpp:1858]: I will add a doc comment.
For example, for `class C { <HERE> }`, `isCompletionDeclContextLocalContext` and `isCodeCompletionAtTopLevel` are both `false`.
I don't think there's a `true`/`true` case though.

[the remark this answers is not in the document]
Comment (benlangmuir) [reply lib/IDE/CodeCompletion.cpp:1858]: Ah, of course, thanks for explaining!

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Review by benlangmuir — APPROVED** [review]

**Pull request merged**